arangod/Aql/RemoteExecutor.h

```diff
@@ -135,6 +135,7 @@ class ExecutionBlockImpl<RemoteExecutor> : public ExecutionBlock {
   /// @brief Mutex to cover against the race, that a getSome request
   ///        is responded before the ticket id is registered.
   arangodb::Mutex _communicationMutex;
+  std::atomic<std::thread::id> _communicationMutexOwner; // current thread owning '_communicationMutex' lock (workaround for non-recusrive MutexLocker)
```

Comment:
OK this is a "fix" i need to reject.

This mutex is used to protect the local state and every place where this mutex is in use cannot be executed (even not using the same thread) while we are in an other place protected by this mutex.

Example:
in getSome we send a message over the ClusterComm.
On return of the ClusterComm (dispatching an asynchrounous call, which allows this code to continue and if the async call is faster it will be held up by exactly this mutex)
Now back the clusterComm will return an OperationID.
This is now retained and we only accept responses for exactly this OperationID.

If now the response is actually injected within the thread and while holding this Mutex (which is the case you run into in your test) the OperationID returned above is NOT yet set.
So the response needs to be rejected (assuming the Mocks are working as intendent) and you cannot get this to work properly.
If the Mocks are not honoring the OperationIDs (which i suspect right now) the mock will break logic depending on this operation IDs in other places.
If we use this RECURSIVE_MUTEX in this code here (even if we just use it in tests) we will NOT see if we get into invalid state, as getting this mutex twice in the thread is disallowed!
We need to see this violation during test phase so we need to keep it a standard mutex.

Reply:
```
> OK this is a "fix" i need to reject.
how do you recommend to test the functionality currently covered by the test?
or should the functionality not have test coverage?
> If now the response is actually injected within the thread and while holding this Mutex (which is the case you run into in your test) the OperationID returned above is NOT yet set.
So the response needs to be rejected (assuming the Mocks are working as intendent) 
I expect the responses generated by the Mock to be returned to the caller
according to the current code of RemoteExecutor there is no reason to reject the responses if their OperationIDs are what is expected
> If the Mocks are not honoring the OperationIDs
the Mocks and the test data are what explicitly set the OperationIDs, there is nothing to not-honour
> If we use this RECURSIVE_MUTEX in this code here (even if we just use it in tests) we will NOT see if we get into invalid state,
that is why I asked you initially if it was ever intended from regular runtime callflow to be possible to call the two methods from the same thread
you told me "if this would happen in production it is broken, so it cannot call it in the same thread"
```